# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter("ignore")

Zet het bestand go_sales_train.sqlite in een makkelijk te vinden map

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales_train.sqlite te maken.

In [2]:
sales_conn = sqlite3.connect('../../go_sales_train.sqlite')

Lees van de ingelezen go_sales_train-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [3]:


tables = ['product', 'product_type', 'product_line', 'sales_staff', 'sales_branch', 'retailer_site', 'country', 'order_header', 'order_details', 'order_method', 'returned_item', 'return_reason']

dataframes = {table: pd.read_sql_query(f"SELECT * FROM {table}", sales_conn) for table in tables}

product = dataframes['product']
product_type = dataframes['product_type']
product_line = dataframes['product_line']
sales_staff = dataframes['sales_staff']
sales_branch = dataframes['sales_branch']
retailer_site = dataframes['retailer_site']
country = dataframes['country']
order_header = dataframes['order_header']
order_details = dataframes['order_details']
returned_item = dataframes['returned_item']
return_reason = dataframes['return_reason']
order_method = dataframes['order_method']

sales_staff



SALES_STAFF_CODE FIRST_NAME  LAST_NAME                   POSITION_EN  \
0                   4      Denis       Pagé                Branch Manager   
1                   5  Élizabeth     Michel  Level 3 Sales Representative   
2                   6      Émile   Clermont  Level 1 Sales Representative   
3                   7    Étienne     Jauvin  Level 2 Sales Representative   
4                  12    Elsbeth  Wiesinger  Level 2 Sales Representative   
..                ...        ...        ...                           ...   
97                120      Giele   Laermans  Level 1 Sales Representative   
98                121   François    De Crée  Level 1 Sales Representative   
99                122     Yvette    Lattrez  Level 3 Sales Representative   
100               123      Willi  Seefelder  Level 2 Sales Representative   
101               124     Sabine     Grüner  Level 3 Sales Representative   

            WORK_PHONE  EXTENSION                FAX                   EMAIL  \
0    +33 1 68 94 52 20      325.0  +33 1 68 94 56 60       DPage@grtd123.com   
1    +33 1 68 94 52 20      336.0  +33 1 68 94 56 60     EMichel@grtd123.com   
2    +33 1 68 94 52 20      378.0  +33 1 68 94 56 60   EClermont@grtd123.com   
3    +33 1 68 94 52 20      398.0  +33 1 68 94 56 60     EJauvin@grtd123.com   
4    +(49) 40 663 1990     1818.0  +(49) 40 663 4571  EWiesinger@grtd123.com   
..                 ...        ...                ...                     ...   
97     +32 16 20.73.21     1340.0    +32 16 20.73.32   GLaermans@grtd123.com   
98     +32 16 20.73.21     1642.0    +32 16 20.73.32     FDecree@grtd123.com   
99     +32 16 20.73.21     1633.0    +32 16 20.73.32    YLattrez@grtd123.com   
100  +(43) 13 79 56 32      325.0  +(43) 13 79 56 33  WSeefelder@grtd123.com   
101  +(43) 13 79 56 32      348.0  +(43) 13 79 56 33     SGruner@grtd123.com   

     DATE_HIRED  SALES_BRANCH_CODE  
0    1996-11-03                  6  
1    1995-06-08                  6  
2    1998-04-07                  6  
3    1997-08-16                  6  
4    1997-02-27                 13  
..          ...                ...  
97   1999-11-15                 38  
98   1999-12-01                 38  
99   1999-12-09                 38  
100  1998-10-28                 39  
101  1998-02-18                 39  

[102 rows x 10 columns]

Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt. <u>Let op:</u> lees eventueel de informatie uit het Notebook van werkcollege 1.1b nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [17]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, sales_conn)
#Op de puntjes hoort de connectie naar go_sales_train óf go_staff_train óf go_crm_train te staan.

name
0         country
1   order_details
2    order_header
3    order_method
4         product
5    product_line
6    product_type
7   retailer_site
8   return_reason
9   returned_item
10   sales_branch
11    sales_staff

erachter 

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [39]:
product_Lhonderd_Hvijftig = (product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100)

resultaat = product.loc[(product_Lhonderd_Hvijftig), ('PRODUCT_NUMBER', 'PRODUCTION_COST')]
resultaat

PRODUCT_NUMBER  PRODUCTION_COST
5                 6            85.11
7                 8            75.00
16               17            60.00
17               18            86.00
19               20            78.55
22               23            65.33
23               24            62.50
45               46            52.86
47               48            55.00
57               58            57.13
68               69            73.33
69               70            80.00
70               71            60.00
71               72            72.50
74               75            80.00
77               78            79.19
78               79            92.58
79               80            94.12
82               83            56.25
83               84            78.55
109             110            60.00
110             111            81.80
113             114            80.00

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [40]:
product_margin_Ltwee_Hzes = (product['MARGIN'] > 0.6) | (product['MARGIN'] < 0.2)

resultaat = product.loc[(product_margin_Ltwee_Hzes), ('PRODUCT_NUMBER', 'MARGIN')]
resultaat

PRODUCT_NUMBER  MARGIN
20              21    0.17
23              24    0.17
85              86    0.67
86              87    0.67
87              88    0.70
88              89    0.63
89              90    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [54]:

country_francs = country['CURRENCY_NAME'] == 'francs'

resultaat = country.loc[(country_francs), ['COUNTRY']]
resultaat

COUNTRY
0        France
7   Switzerland
15      Belgium

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [53]:
# sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
#
# pd.read_sql(sql_query, sales_conn)

product_margin_Hvijftig = product['MARGIN'] > 0.5

resultaat = product.loc[(product_margin_Hvijftig), ['INTRODUCTION_DATE']].drop_duplicates('INTRODUCTION_DATE')
resultaat

INTRODUCTION_DATE
66         2000-10-26
85         1995-02-15
101        2003-12-10
103        2003-12-18
104        2003-12-27
106        2004-01-13
110        2003-12-15

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [65]:
sales_branch_adres_twee_notNan = sales_branch['ADDRESS2'].notna() & sales_branch['REGION'].notna()

resultaat = sales_branch.loc[(sales_branch_adres_twee_notNan), ['ADDRESS1', 'CITY']]
resultaat




ADDRESS1              CITY
2                Singelgravenplein 4         Amsterdam
13  Prol. Paseo de la Reforma No. 51  Distrito Federal
14               202-2-3 Hyakunincho             Tokyo
15                     543-225 Asahi        Osaka City
16                  2315 Queen's Ave         Melbourne
18             Avenida Paulista, 333         São Paulo
24                    3 Albany Court        Birmingham

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [73]:

country_with_dollars = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')

resultaat = country.loc[(country_with_dollars), ['COUNTRY']].sort_values('COUNTRY')
resultaat

COUNTRY
14      Australia
3          Canada
11         Taiwan
2   United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [88]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

klanten_uit_D = (retailer_site['POSTAL_ZONE'].str[0] == 'D') & (retailer_site['ADDRESS2'].notna())

resultaat = retailer_site.loc[(klanten_uit_D), ['ADDRESS1', 'ADDRESS2', 'CITY']]
resultaat





ADDRESS1       ADDRESS2       CITY
31  Röntgenstraße 90  3. Tür rechts  Frankfurt
59   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [94]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

returned_aantal_items = returned_item['RETURN_QUANTITY'].sum()

print(returned_aantal_items)

11908


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [101]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

aantal_region_sales_branch = sales_branch['REGION'].nunique()



print(aantal_region_sales_branch)


15


Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [114]:
product_H = product['MARGIN'].max()
product_L = product['MARGIN'].min()
product_Average = product['MARGIN'].mean().round(2)

df_resultaat = pd.DataFrame({'Hoogste_Margin' : [product_H], 'Laagste_Margin' : [product_L], "Average_Margin" : [product_Average]})


df_resultaat

Hoogste_Margin  Laagste_Margin  Average_Margin
0             0.7            0.17             0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [140]:
sql_query = "SELECT name FROM sqlite_master WHERE type = 'table';"
pd.read_sql(sql_query, sales_conn)

aantal_Klanten_vestigingen_zonderTweedeAdres = retailer_site['ADDRESS2'].isna()

resultaat = retailer_site.loc[(aantal_Klanten_vestigingen_zonderTweedeAdres), :].shape[0]
print(resultaat)




286


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [152]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

gem_kostprijs_verkochte_producten = (order_details['UNIT_SALE_PRICE']) < (order_details['UNIT_PRICE'])

nieuw_order_details = order_details.loc[(gem_kostprijs_verkochte_producten), :]

resultaat = nieuw_order_details['UNIT_SALE_PRICE'].mean().round(2)

print(resultaat)


180.44


Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [159]:


aantal_staff_per_functie = sales_staff.groupby('POSITION_EN', as_index = False)['SALES_STAFF_CODE'].nunique().rename(columns = {'SALES_STAFF_CODE':'aantal_medewerker'})


aantal_staff_per_functie


POSITION_EN  aantal_medewerker
0                Branch Manager                 19
1              District Manager                  4
2               General Manager                  1
3  Level 1 Sales Representative                 13
4  Level 2 Sales Representative                 39
5  Level 3 Sales Representative                 24
6              Regional Manager                  2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [169]:


tele_Met_Medewerkers = sales_staff.groupby('WORK_PHONE', as_index = False)['SALES_STAFF_CODE'].nunique().rename(columns = {'SALES_STAFF_CODE':'aantal_medewerker'})


tele_Met_Vier_Medewerkers = tele_Met_Medewerkers.loc[(tele_Met_Medewerkers['aantal_medewerker'] > 4), :]

tele_Met_Vier_Medewerkers

WORK_PHONE  aantal_medewerker
4     +(44) 121 3505267                  5
10  +(61) 03 2982  4242                  5
11  +31 (0)20 692 93 94                  5
13    +33 1 68 94 52 20                  5
18     1 (206) 292-0012                  5
19     1 (305) 557-4810                  6
20     1 (310) 281-5722                  5
21     1 (403) 232-5986                  5
22     1 (416) 493-5595                  5
23     1 (617) 268-6754                  6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [180]:


retailer_site_country_merge = pd.merge(retailer_site, country, left_on = 'COUNTRY_CODE', how = 'inner', right_on = 'COUNTRY_CODE')

adres_stad_Nederland = retailer_site_country_merge.loc[(retailer_site_country_merge['COUNTRY'] == 'Netherlands'), ['ADDRESS1', 'CITY']]

adres_stad_Nederland





ADDRESS1          CITY
240                   Beets 36         Beets
241               Perestraat 1      Deventer
242  Laan van Meerdervoort 966      Den Haag
243             Kalfjeslaan 47    Amstelveen
244               Startbaan 11    Amstelveen
245      Noorderspoorsingel 28     Groningen
246              Hogehilweg 13     Amsterdam
247             Spoorstraat 43    Varsseveld
248             Westhavenweg 8     Amsterdam
249              Westzeedijk 4        Muiden
250            Gooiergracht 91         Laren
251            Hoofdstraat 150   Pieterburen
252                Olde Diek 2         Laren
253     Amsterdamsestraatweg 8      Maarssen
254        Krayenhoffstraat 11     Amsterdam
255         Overschiestraat 19     Amsterdam
256              Havenbuurt 12        Marken
257               Duivendaal 1    Wageningen
258              Voorstraat 37  Nieuweschans
259         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen)

In [187]:

product_producttype_merge = pd.merge(product, product_type, on = 'PRODUCT_TYPE_CODE', how = 'inner')

product_van_type_eyewear = product_producttype_merge.loc[(product_producttype_merge['PRODUCT_TYPE_EN'] == 'Eyewear'), 'PRODUCT_NAME']

product_van_type_eyewear

67        Polar Sun
68        Polar Ice
69     Polar Sports
70       Polar Wave
71    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [245]:

sales_branch_staff_merge = pd.merge(sales_staff, sales_branch, on = 'SALES_BRANCH_CODE', how = 'inner')

sales_branch_manager = sales_branch_staff_merge.loc[(sales_branch_staff_merge['POSITION_EN'] == 'Branch Manager'), :]



order_header_retailer_merge = pd.merge(retailer_site, order_header, on = 'RETAILER_SITE_CODE', how = 'inner')

order_header_retailer_merge = order_header_retailer_merge.drop_duplicates(subset=['ADDRESS1'])

final_merge = pd.merge(sales_branch_manager,order_header_retailer_merge, on = 'SALES_BRANCH_CODE', how = 'inner', )

result = final_merge[['ADDRESS2_x', 'FIRST_NAME', 'LAST_NAME']]

nt = result['ADDRESS2_x'].notna()

resultaat = result.loc[nt, :]

resultaat


#FUCK DIT KUT OPGAVE




ADDRESS2_x FIRST_NAME LAST_NAME
61   4e verdieping       Kick   Kalkman
62   4e verdieping       Kick   Kalkman
63   4e verdieping       Kick   Kalkman
64   4e verdieping       Kick   Kalkman
65   4e verdieping       Kick   Kalkman
..             ...        ...       ...
284  Market Street      Polly    Linton
285  Market Street      Polly    Linton
286  Market Street      Polly    Linton
287  Market Street      Polly    Linton
288  Market Street      Polly    Linton

[66 rows x 3 columns]

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

In [36]:
rsql_query = "SELECT name FROM sqlite_master WHERE type='table';"



sales_orders_merge = pd.merge(sales_staff, order_header, on = 'SALES_STAFF_CODE', how = 'left')

manager = sales_orders_merge[sales_orders_merge['POSITION_EN'].str.contains('Manager', case= False, na=False)]

result = manager[['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

result





POSITION_EN  ORDER_DATE
0       Branch Manager         NaN
320   Regional Manager         NaN
648   District Manager         NaN
3748    Branch Manager  2024-09-09
3749    Branch Manager  2024-03-13
3750    Branch Manager  2024-01-09
4186   General Manager         NaN

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [49]:

product_producttype_merge = pd.merge(product, product_type, on = 'PRODUCT_TYPE_CODE', how = 'inner' )

product_order_merge = pd.merge(product_producttype_merge, order_details, on = 'PRODUCT_NUMBER', how = 'inner')

zeven_vijftig_plus_orders = product_order_merge[product_order_merge['QUANTITY'] > 750]

result = zeven_vijftig_plus_orders[['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates('PRODUCT_NAME')
result

PRODUCT_NAME    PRODUCT_TYPE_EN
5427                Star Peg              Tents
11452              Firefly 2           Lanterns
28608        BugShield Spray  Insect Repellents
28892  BugShield Lotion Lite  Insect Repellents
29430       BugShield Lotion  Insect Repellents
29442      BugShield Extreme  Insect Repellents
30741         Sun Shelter 15          Sunscreen
31164         Sun Shelter 30          Sunscreen
31775             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [57]:

product_order_header_merge = pd.merge(product, order_details, on = 'PRODUCT_NUMBER', how = 'inner')

veertig_korting = product_order_header_merge[((product_order_header_merge['UNIT_PRICE'] - product_order_header_merge['UNIT_SALE_PRICE']) / product_order_header_merge['UNIT_PRICE']) > 0.4]

result = veertig_korting[['PRODUCT_NAME']].drop_duplicates()

result

PRODUCT_NAME
27830             BugShield Natural
28283               BugShield Spray
28694         BugShield Lotion Lite
29127              BugShield Lotion
29464             BugShield Extreme
30562                Sun Shelter 15
30929                Sun Shelter 30
34441  Hailstorm Titanium Woods Set

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [69]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

odtr_merge = pd.merge(order_details, returned_item, on = 'ORDER_DETAIL_CODE', how = 'inner')

odtr_reason_merge = pd.merge(odtr_merge, return_reason, on = 'RETURN_REASON_CODE', how = 'inner')

odtr_final = odtr_reason_merge[(odtr_reason_merge['RETURN_QUANTITY']/odtr_reason_merge['QUANTITY']) >  0.9]

result = odtr_final[['RETURN_DESCRIPTION_EN']].drop_duplicates()

result

RETURN_DESCRIPTION_EN
0   Wrong product ordered
4   Wrong product shipped
5  Unsatisfactory product

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [87]:
product_producttype_merge = pd.merge(product_type, product, on = 'PRODUCT_TYPE_CODE', how = 'left')

result = product_producttype_merge.groupby('PRODUCT_TYPE_EN', as_index= False)['PRODUCT_NUMBER'].count(). rename(columns = {'PRODUCT_NUMBER':'Aantal_producten'})

result

PRODUCT_TYPE_EN  Aantal_producten
0             Binoculars                 4
1   Climbing Accessories                 7
2           Cooking Gear                10
3                Eyewear                 5
4              First Aid                 5
5       Golf Accessories                 4
6      Insect Repellents                 5
7                  Irons                 4
8                 Knives                 5
9               Lanterns                12
10            Navigation                 4
11                 Packs                 6
12               Putters                 3
13                  Rope                 4
14                Safety                 4
15         Sleeping Bags                 7
16             Sunscreen                 5
17                 Tents                 6
18                 Tools                 6
19               Watches                 5
20                 Woods                 4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [88]:
retailer_site_country_merge = pd.merge(country, retailer_site, on = 'COUNTRY_CODE', how = 'left')

result = retailer_site_country_merge.groupby('COUNTRY', as_index = False)['RETAILER_SITE_CODE'].count().rename(columns = {'RETAILER_SITE_CODE':'Aantal_vestigingen'})

result

COUNTRY  Aantal_vestigingen
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  31
5            China                  10
6          Denmark                   6
7          Finland                   5
8           France                  29
9          Germany                  32
10           Italy                  12
11           Japan                  25
12           Korea                  10
13          Mexico                  10
14     Netherlands                  20
15           Spain                   9
16          Sweden                  12
17     Switzerland                  13
18          Taiwan                  10
19  United Kingdom                  25
20   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [101]:
product_producttype_merge = pd.merge(product_type, product, on = 'PRODUCT_TYPE_CODE', how = 'right')

cooking_gear_product = product_producttype_merge[product_producttype_merge['PRODUCT_TYPE_EN'] == 'Cooking Gear']

last_merger = pd.merge(cooking_gear_product, order_details, on = 'PRODUCT_NUMBER', how = 'inner')

result = last_merger.groupby('PRODUCT_NAME', as_index = False).agg({'QUANTITY':'sum', 'UNIT_SALE_PRICE':'mean'}).round({'UNIT_SALE_PRICE' : 2}).sort_values('QUANTITY')

result


PRODUCT_NAME  QUANTITY  UNIT_SALE_PRICE
3  TrailChef Deluxe Cook Set      3110           121.22
7     TrailChef Single Flame      8504            63.83
5           TrailChef Kettle      8826            12.20
1         TrailChef Cook Set     10878            51.69
4     TrailChef Double Flame     12476           132.67
8         TrailChef Utensils     13776            17.29
6      TrailChef Kitchen Kit     22206            22.35
0          TrailChef Canteen     26544            11.98
2              TrailChef Cup     27418             6.98
9        TrailChef Water Bag     36738             6.03

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [112]:

country_sales_branch_merge = pd.merge(country, sales_branch, on = 'COUNTRY_CODE', how = 'left')

country_retail_merge = pd.merge(country_sales_branch_merge, retailer_site, on = 'COUNTRY_CODE', how = 'inner')


result = country_retail_merge.groupby(['COUNTRY', 'CITY_x'], as_index = False)['CITY_y'].count().rename(columns = {'CITY_x':'verkoper', 'CITY_y' : 'klanten'})

result






COUNTRY          verkoper  klanten
0        Australia         Melbourne       10
1          Austria              Wien       10
2          Belgium          Heverlee       10
3           Brazil         São Paulo        1
4           Canada           Calgary       31
5           Canada           Toronto       31
6            China          Shanghai       10
7          Finland            Kuopio        5
8           France              Lyon       29
9           France             Paris       29
10         Germany           Hamburg       32
11         Germany           München       32
12           Italy            Milano       12
13           Japan        Osaka City       25
14           Japan             Tokyo       25
15           Korea             Seoul       10
16          Mexico  Distrito Federal       10
17     Netherlands         Amsterdam       20
18           Spain            Bilbao        9
19          Sweden             Kista       12
20     Switzerland            Zürich       13
21          Taiwan         San Chung       10
22  United Kingdom        Birmingham       25
23  United Kingdom            London       25
24   United States            Boston      101
25   United States       Los Angeles      101
26   United States             Miami      101
27   United States           Seattle      101

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [121]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
pd.read_sql(sql_query, sales_conn)

# merger = pd.merge(sales_staff, order_header, on = 'SALES_STAFF_CODE', how = 'outer')
#
# result = merger.loc[merger['ORDER_NUMBER'].isna(), ['FIRST_NAME', 'LAST_NAME']]



verkopers = order_header['SALES_STAFF_CODE'].dropna().unique()

result = sales_staff.loc[~sales_staff['SALES_STAFF_CODE'].isin(verkopers), ['FIRST_NAME', 'LAST_NAME']]

result

FIRST_NAME      LAST_NAME
0       Denis           Pagé
5        Else         Mörike
6       Frank          Fuchs
9       Fritz         Hirsch
10       Jörg          Kunze
12      Maria      Iacobucci
17       Kick        Kalkman
22      Karin      Bergström
25      Sally          White
27      Frank        Bretton
28    Georges  Saint-Germain
33       Alex      Rodriguez
37   Samantha          Floyd
41      Julie          Olsen
45      Corey         Wright
51    Estelle        Johnson
52       Greg        Belding
55      Julie        Béchard
59      Lucía          Reyna
64     Kazumi        Uragome
66      Alice         Walter
84     Walter         Taylor
85   Penelope        Tamrine
88      Polly         Linton
96      Gracy        Gellens

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [130]:


average_margin = product['MARGIN'].mean()
filtered_df = product[product['MARGIN'] < average_margin]

result = pd.DataFrame({
    'Aantal producten' : [filtered_df.shape[0]],
    'Gemiddelede marge' : [filtered_df['MARGIN'].mean().round(2)]
})
result


Aantal producten  Gemiddelede marge
0                59               0.29

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [159]:


order_vijfhonderd = order_details[order_details['UNIT_SALE_PRICE'] > 500]


all_returned = returned_item['ORDER_DETAIL_CODE'].dropna().unique()

never_returned = order_vijfhonderd.loc[~order_vijfhonderd['ORDER_DETAIL_CODE'].isin(all_returned), ['PRODUCT_NUMBER']]



result = product.loc[product['PRODUCT_NUMBER'].isin(never_returned.squeeze()), ['PRODUCT_NAME']]

result


PRODUCT_NAME
11                           Star Dome
12                        Star Gazer 2
13                        Star Gazer 3
14                        Star Gazer 6
44                      Husky Rope 200
100              Hailstorm Steel Irons
101           Hailstorm Titanium Irons
102         Lady Hailstorm Steel Irons
103      Lady Hailstorm Titanium Irons
104       Hailstorm Titanium Woods Set
105          Hailstorm Steel Woods Set
106  Lady Hailstorm Titanium Woods Set
107     Lady Hailstorm Steel Woods Set

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [8]:

medewerkers_data = sales_staff.loc[:, ["LAST_NAME", "POSITION_EN"]]

for index, medewerker in medewerkers_data.iterrows():
    rol = medewerkers_data.at[index, 'POSITION_EN']
    if "Manager" in rol:
        medewerkers_data.at[index, 'IS-MANAGER'] = "Ja"
    else:
        medewerkers_data.at[index, 'IS-MANAGER'] = "Nee"

medewerkers_data = medewerkers_data.drop("POSITION_EN", axis=1)
medewerkers_data

LAST_NAME IS-MANAGER
0         Pagé         Ja
1       Michel        Nee
2     Clermont        Nee
3       Jauvin        Nee
4    Wiesinger        Nee
..         ...        ...
97    Laermans        Nee
98     De Crée        Nee
99     Lattrez        Nee
100  Seefelder        Nee
101     Grüner        Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [10]:
from datetime import date
date.today().year

2025

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [11]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [12]:
medewerkers_data = sales_staff.loc[:, ["LAST_NAME", "DATE_HIRED"]]

huidig_jaar = date.today().year

for index, medewerker in medewerkers_data.iterrows():
    date_hired = medewerkers_data.at[index, 'DATE_HIRED']

    date_str = date_hired
    date_format = '%Y-%m-%d'
    date_obj = datetime.strptime(date_str, date_format)

    aantal_jaar_in_dienst = huidig_jaar - date_obj.year


    if aantal_jaar_in_dienst < 25:
        medewerkers_data.at[index, 'DIENSTVERBAND'] = "Kort in dienst"
    else:
        medewerkers_data.at[index, 'DIENSTVERBAND'] = "Lang in dienst"

medewerkers_data = medewerkers_data.drop("DATE_HIRED", axis=1)
medewerkers_data

LAST_NAME   DIENSTVERBAND
0         Pagé  Lang in dienst
1       Michel  Lang in dienst
2     Clermont  Lang in dienst
3       Jauvin  Lang in dienst
4    Wiesinger  Lang in dienst
..         ...             ...
97    Laermans  Lang in dienst
98     De Crée  Lang in dienst
99     Lattrez  Lang in dienst
100  Seefelder  Lang in dienst
101     Grüner  Lang in dienst

[102 rows x 2 columns]

## Van Jupyter Notebook naar Pythonproject

1. Richt de map waarin jullie tot nu toe hebben gewerkt in volgens de mappenstructuur uit de slides.
2. Maak van de ontstane mappenstructuur een Pythonproject dat uitvoerbaar is vanuit de terminal. Maak daarin een .py-bestand dat minstens 5 antwoorden uit dit notebook (in de vorm van een DataFrame) exporteert naar Excelbestanden. Alle notebooks mogen als notebook blijven bestaan.
3. Zorg ervoor dat dit Pythonproject zijn eigen repo heeft op Github. Let op: je virtual environment moet <b><u>niet</u></b> meegaan naar Github.

Je mag tijdens dit proces je uit stap 1 ontstane mappenstructuur aanpassen, zolang je bij het beoordelingsmoment kan verantwoorden wat de motivatie hierachter is. De slides verplichten je dus nergens toe.